### PR TITLE
DRBG: delay initialization of DRBG method until instantiation

### PR DESCRIPTION
**Pull request by mspncp**
> ~Could not resist to reserve this beautiful number for my next pull request ;-)~

Here it comes: This pull request is based on #11010, see https://github.com/openssl/openssl/pull/11010#issuecomment-586947393 for a detailed explanation of what it does. I had to rebase @slontis' pull request in order to pick up @levitte's fix f93a17f13cef29ec73fc25aa186d240b439963f8. Only the last two commits are relevant:

commit f428aec8bd1094a7bc0ca95bb99750fe3be7eba4

    Check that the DRBG's internal state has been zeroized after uninstantiation

commit bc826c0f35c1176314c31e9fbd4fb68305b13863

    DRBG: delay initialization of DRBG method until instantiation
    
    Previously, the initialization was done immediately in RAND_DRBG_set(),
    which is also called in RAND_DRBG_uninstantiate().
    
    This made it difficult for the FIPS DRBG self test to verify that the
    internal state had been zeroized, because it had the side effect that
    the drbg->data structure was reinitialized immediately.
    
    To solve the problem, RAND_DRBG_set() has been split in two parts
    
        static int rand_drbg_set(RAND_DRBG *drbg, int type, unsigned int flags);
        static int rand_drbg_init_method(RAND_DRBG *drbg);
    
    and only the first part is called from RAND_DRBG_uninstantiate().

The pull request is marked WIP, because the test_rand is still failing. The internal logic of the delay initialization still needs some fine-tuning.

### CODE REVIEW

**Comment by mattcaswell** [comment]
I note by the way that 11,111 is a product of 2 primes: 41 * 271

**Comment by mspncp** [comment]
That's not enough for RSA, right?

**Comment by mspncp** [comment]
While we're at it:

* https://www.numberempire.com/11111
* https://numberworld.info/11111
* http://www.numbersaplenty.com/11111
* https://math.stackexchange.com/questions/1591010/property-of-111-111

**Comment by mattcaswell** [comment]
> That's not enough for RSA, right?

Not to do PKCS1 padding....not sure about "no padding". I think its safe to say I wouldn't trust my bank details with it.

**Comment by DDvO** [comment]
I'm tempted to (ab-)use this PR as a degenerate test case for the latest `ghmerge` of https://github.com/openssl/tools/pull/59

**Comment by mspncp** [comment]
> I'm tempted to (ab-)use this PR as a degenerate test case for the latest `ghmerge` of [openssl/tools#59](https://github.com/openssl/tools/pull/59)

Sorry for spoiling your fun. I told you, it was only a placeholder for my next pull request ;-)

**Comment by DDvO** [comment]
> > I'm tempted to (ab-)use this PR as a degenerate test case for the latest `ghmerge` of [openssl/tools#59](https://github.com/openssl/tools/pull/59)
> 
> Sorry for spoiling your fun. I told you, it was only a placeholder for my next pull request ;-)

What a shame, I was already planning to approve the empty list of commits since I was sure it was error-free :rofl:

**Comment by paulidale** [comment]
> I note by the way that 11,111 is a product of 2 primes: 41 * 271

I don't think we'll be around when the next [repunit prime](https://en.wikipedia.org/wiki/Repunit#Repunit_primes) is reached.

File: providers/fips/self_test_kats.c [hunk 1]
```diff
@@ -193,6 +198,108 @@ static int self_test_kdf(const ST_KAT_KDF *t, OSSL_ST_EVENT *event,
...
+{
+    int ret = 0;
+    unsigned char out[256];
+    RAND_DRBG *drbg = NULL;
+    unsigned int flags = 0;
+    int prediction_resistance = 1; /* Causes a reseed */
+    OSSL_PARAM drbg_params[3] = {
+        OSSL_PARAM_END, OSSL_PARAM_END, OSSL_PARAM_END
+    };
+    static const unsigned char zero[sizeof(drbg->data)] = { 0 };
```
Comment (slontis) [inline providers/fips/self_test_kats.c]: This looks like a NULL access of drbg?

Comment (paulidale) [reply providers/fips/self_test_kats.c]: It isn't.

Comment (mspncp) [reply providers/fips/self_test_kats.c]: As Pauli already noted: it happens at compile time. So unless the compiler is buggy...

**Comment by slontis** [comment]
Should this be rebased after my PR goes in?

**Comment by mspncp** [comment]
Yes, that was the intention.

**Comment by mspncp** [comment]
Rebased and squashed (without further changes) onto 980a880ee5a85b94caac3049c7d3be0b765d97d3 (#11010).

**Comment by mspncp** [comment]
Ready for review.

**Comment by levitte** [comment]
Ready to merge!

**Comment by mspncp** [comment]
Merged, thanks for the review!

e70452155e Check that the DRBG's internal state has been zeroized after uninstantiation
75ff4f7404 DRBG: delay initialization of DRBG method until instantiation

**Comment by slontis** [comment]
Thanks for doing this @mspncp

**Comment by mspncp** [comment]
You're welcome. Always a pleasure :-).

**Pull request closed without merging**